File: backend/app/schemas/pipeline.py
from pydantic import BaseModel
from typing import Optional


class TranscribeRequest(BaseModel):
    language: Optional[str] = None  # auto-detect if None


class TranscribeSegment(BaseModel):
    id: str
    start_time: float
    end_time: float
    text: str


class TranscribeResponse(BaseModel):
    segments: list[TranscribeSegment]
    detected_language: str


class TranslateRequest(BaseModel):
    segments: list[TranscribeSegment]
    source_language: str
    target_language: str


class TranslatedSegment(BaseModel):
    id: str
    start_time: float
    end_time: float
    original_text: str
    translated_text: str


class TranslateResponse(BaseModel):
    segments: list[TranslatedSegment]
    source_language: str
    target_language: str


class SynthesizeRequest(BaseModel):
    text: str
    voice_id: str
    language: str
    speed: Optional[float] = 1.0


class VoiceOption(BaseModel):
    voice_id: str
    name: str
    language: str
    gender: str


class VoiceListResponse(BaseModel):
    voices: list[VoiceOption]


class SplitSegmentsRequest(BaseModel):
    segments: list[TranslatedSegment]
    max_duration: float = 2.0
    max_lines: int = 3
    target_language: str


class SplitSegmentsResponse(BaseModel):
    segments: list[TranslatedSegment]


class WordTimestamp(BaseModel):
    word: str
    start: float  # seconds relative to audio start
    end: float    # seconds relative to audio start


class RealignResponse(BaseModel):
    words: list[WordTimestamp]
    duration: float  # total audio duration in seconds

File: functions/src/types/pipeline.ts
export interface TranscribeSegment {
  id: string;
  start_time: number;
  end_time: number;
  text: string;
}

export interface TranscribeResponse {
  segments: TranscribeSegment[];
  detected_language: string;
}

export interface TranslateRequest {
  segments: TranscribeSegment[];
  source_language: string;
  target_language: string;
}

export interface TranslatedSegment {
  id: string;
  start_time: number;
  end_time: number;
  original_text: string;
  translated_text: string;
}

export interface TranslateResponse {
  segments: TranslatedSegment[];
  source_language: string;
  target_language: string;
}

export interface SynthesizeRequest {
  text: string;
  voice_id: string;
  language: string;
  speed?: number; // default 1.0
}

// NOTE: SynthesizeResponse is { audioBase64: string } — NOT raw binary
// This is different from Python backend which returned raw bytes
export interface SynthesizeResponse {
  audioBase64: string;
}

export interface VoiceOption {
  voice_id: string;
  name: string;
  language: string;
  gender: string;
}

export interface VoiceListResponse {
  voices: VoiceOption[];
}

export interface SplitSegmentsRequest {
  segments: TranslatedSegment[];
  max_duration?: number; // default 2.0
  max_lines?: number; // default 3
  target_language: string;
}

export interface SplitSegmentsResponse {
  segments: TranslatedSegment[];
}

export interface WordTimestamp {
  word: string;
  start: number; // seconds
  end: number;   // seconds
}

export interface RealignResponse {
  words: WordTimestamp[];
  duration: number; // total audio duration
}

// For Cloud Function calls (base64 audio input)
export interface TranscribeRequest {
  audioBase64: string;
  filename: string;
  language?: string;
}

export interface RealignRequest {
  audioBase64: string;
  filename?: string;
}
